### ISSUE TITLE
response()->json() output lets loaded relations override properties.

### ISSUE THREAD

**Issue by mace015** [laravel/framework#27624]
- Laravel Version: 5.7.27
- PHP Version: 7.2.11
- Database Driver & Version: Not relevant

### Description:
When I load a relation on say the user model (doesn't matter), invoices for example, and then try to modify the users invoices (in my case filter out some unwanted results), the results of the filter are written to the user->invoices property/attribute, but the original relation results remain the same in the model, as expected.

When I try to return the $user object in this case, the JSON output overrides the invoices property/attribute with the previously loaded relation, thus returning all relations results instead of the filtered ones.

I'm not entirely sure if this is an actual bug, but it sure didn't do what I was expecting.

### Steps To Reproduce:

```php
Route::get('/test/{user}', function (User $user) {
    $user->load(['invoices']);

    $user->invoices = null;

    return response()->json([
        'user' => $user
        // the content of $user->invoices after this point is the relations loaded on the load call, not null.
    ]);
});
```

**Comment by staudenmeir** [comment]
`$user->invoices = null;` only affects the model's `$attributes` property, but the relationships are stored in `$relations`.

Use `setRelation()` instead:

```php
$user->setRelation('invoices', null);
```

**Comment by driesvints** [comment]
Can you try the solution proposed above?

**Comment by mace015** [comment]
@staudenmeir @driesvints 
I came to that conclusion as well, as I said in my description of the problem, the behavior on the model itself is as expected, however, when outputting the model as JSON, I would except that the property `invoices` has precedence over the relation `invoices`.

**Comment by mace015** [comment]
@staudenmeir @driesvints 
This originates in `Illuminate\Database\Eloquent\Model:1076` in the `toArray` method.
As you can see the relations are merged into the attributes array, causing the override.

**Comment by driesvints** [comment]
It is because `$this->invoices` only references your actual relationship and not the state your model currently is in. You'll first need to save its state before continuing and/or either refresh it.

**Comment by mace015** [comment]
I don't think this issue should be closed just yet.

If the behavior of `toArray` is as intended, then the order/behaviour of `getAttribute` is inconsistent with `toArray`.
You see, when I dd `$user->invoices` before the response()->json() in my example above, I do get the null value.

This is because in `Illuminate\Database\Eloquent\Concerns\HasAttributes` in `getAttribute`, the existence of the attribute is checked, and if it exists that value is returned, only after that the relations existence is checked and returned if present.

So basically:
In `getAttribute` the attribute `invoices` has precedence (which is logical in my opinion).
In `toArray` the relation `invoices` has precedence.

**Comment by staudenmeir** [comment]
I see your point, but I would argue that there shouldn't be a case where `$attributes` and `$relations` contain the same key. Relationships aren't really meant to be changed like this.
